File: backend/node/api/src/application/use-cases/sync-steps.ts
import type { StepInput } from "../../domain/entities/pipeline";
import { Pipeline } from "../../domain/entities/pipeline";
import { inject } from "../../infra/DI/container";
import type { PipelineRepository } from "../../infra/repository/pipeline";

type Input = {
	pipelineId: string;
	steps: StepInput[];
};

type Output = {
	created: number;
	updated: number;
	deleted: number;
};

export class SyncStepsUseCase {
	@inject("pipeline-repository")
	private declare readonly pipelineRepository: PipelineRepository;

	async execute(
		input: Input,
	): Promise<[Output, undefined] | [undefined, Error]> {
		const [pipeline, pipelineError] = await this.pipelineRepository.getById(
			input.pipelineId,
		);
		if (pipelineError) return [undefined, pipelineError];

		const [chain, restoreError] = Pipeline.restoreSteps(pipeline, input.steps);
		if (restoreError) return [undefined, restoreError];

		const [, setStepsError] = pipeline.setSteps(chain);
		if (setStepsError) return [undefined, setStepsError];

		const [result, syncError] = await this.pipelineRepository.sync(pipeline);
		if (syncError) return [undefined, syncError];

		return [result, undefined];
	}
}

File: backend/go/api/internal/pipeline/use-cases/sync-steps.go
package usecases

import (
	"context"

	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/pipeline"
	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/shared/errors"
)

type SyncStepsUseCaseDependencies struct {
	PipelineRepository *pipeline.PipelineRepository
}

type SyncStepsUseCase struct {
	SyncStepsUseCaseDependencies
}

func NewSyncStepsUseCase(deps SyncStepsUseCaseDependencies) SyncStepsUseCase {
	return SyncStepsUseCase{
		SyncStepsUseCaseDependencies: deps,
	}
}

type StepInput struct {
	ID          string
	Name        *string
	Description *string
	Operation   string
	By          int
	NextStepId  *string
}

type SyncStepsInput struct {
	PipelineId string
	Steps      []StepInput
}

type SyncStepsOutput struct {
	Created int
	Updated int
	Deleted int
}

func (u *SyncStepsUseCase) Execute(input SyncStepsInput) (SyncStepsOutput, error) {
	p, err := u.PipelineRepository.GetByID(context.Background(), input.PipelineId)
	if err != nil {
		return SyncStepsOutput{}, err
	}

	if p == nil {
		return SyncStepsOutput{}, errors.PipelineNotFoundError(input.PipelineId)
	}

	stepInputs := make([]pipeline.StepInput, len(input.Steps))
	for i, s := range input.Steps {
		stepInputs[i] = pipeline.StepInput{
			ID:          s.ID,
			Name:        s.Name,
			Description: s.Description,
			Operation:   s.Operation,
			By:          s.By,
			NextStepId:  s.NextStepId,
		}
	}

	chain, err := pipeline.RestoreSteps(p, stepInputs)
	if err != nil {
		return SyncStepsOutput{}, err
	}

	if err := p.SetSteps(chain); err != nil {
		return SyncStepsOutput{}, err
	}

	result, err := u.PipelineRepository.Sync(context.Background(), p)
	if err != nil {
		return SyncStepsOutput{}, err
	}

	return SyncStepsOutput{
		Created: result.Created,
		Updated: result.Updated,
		Deleted: result.Deleted,
	}, nil
}
